File: script/oca_to_linkml.py
# oca_to_linkml.py
#
# OCA to LinkML via DH schema_core.tsv, schema_slots.tsv and schema_enums.tsv
# tables which can then be used by tabular_to_schema.py to create LinkML / DH
# schema files schema.yaml and schema.json, as well as language variants.
#
# Author: Damion Dooley
#
# Run this in folder where one wants DataHarmonizer schema to be built, so 
# typically, in some DH web/templates/[schema folder]/, which reads by default
# a file called oca_bundle.json
#
# > python3 ../../../script/oca_to_linkml.py [-i oca_bundle_file_name]
#
# Output: 
#
# 	schema_core.yaml 	- the main LinkML upper yaml structure .
#	schema_slots.tsv 	- the list of column fields added to above yaml.
#	schema_enums.tsv 	- the list of enumerations (code lists) added to above.
#
# Then in DH context, run "> python3 tabular_to_schema.py" to generate
#
# 	schema.yaml	- yaml version of complete LinkML schema for default language
#	schema.json - json version of schema.yaml used by DataHarmonizer
#
# Also for any language locale variants, a language overlay file which is 
# layered on top of the above schema.json file:
#
# 	locales/[i18n locale code]/schema.yaml
# 	locales/[i18n locale code]/schema.json
#
# DataHarmonizer can load the complete schema.json file directly via
# "load template" option.  However access to multilingual functionality will
# require adding the complete schema into the schema bundle and menu.js file.
#
# Detecting OCA data types via regular expression Numeric, Text, 
#
# Numeric:
# integer or decimal number, may begin with + or -	/^[-+]?\d*\.?\d+$/gm
# integer		/^-?[0-9]+$/gm
# 
# Textual:
# Entries of any length with only capital letters		^[A-Z]*$
# Capital or lower case letters only, at least 1 character, and 50 characters max		^[A-Za-z]{1,50}$
# Capital or lower case letters only, 50 characters max		^[A-Za-z]{0,50}$
# Short text, 50 characters max		^.{0,50}$
# Short text, 250 characters max		^.{0,250}$
# long text, 800 characters max		^.{0,800}$
# long text, 4000 characters max		^.{0,4000}$
# Canadian postal codes (A1A 1A1)		^[A-Z][0-9][A-Z]\s[0-9][A-Z][0-9]$
# Zip code		^\d{5,6}(?:[-\s]\d{4})?$
# Email address		[a-zA-Z0-9_\.\+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-\.]+
# URL	https?\:\/\/[a-zA-Z0-9\-\.]+\.[a-zA-Z]{2,}
# Phone number		\+?\(?\d{2,4}\)?[\d\s-]{3,}
# Latitude in formats S30°15'45.678" or N12°30.999"		^[NS]-?(?:[0-8]?\d|90)°(?:\d+(?:\.\d+)?)(?:'(\d+(?:\.\d+)?)")?$
# Longitude in formats E30°15'45.678" or W90°00.000"		^[WE]-?(?:[0-8]?\d|90)°(?:\d+(?:\.\d+)?)(?:'(\d+(?:\.\d+)?)")?$
#
# Date
#
# ISO: YYYY-MM-DD: year month day	^(\d{4})-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01])$
# ISO: YYYYMMDD: year month day	^(\d{4})(0[1-9]|1[0-2])(0[1-9]|[1-2]\d|3[0-1])$
# ISO: YYYY-MM: year month	^(\d{4})-(0[1-9]|1[0-2])$
# ISO: YYYY-Www: year week (e.g. W01)	^(?:\d{4})-W(0[1-9]|[1-4][0-9]|5[0-3])$
# ISO: YYYYWww: year week (e.g. W01)	^(?:\d{4})W(0[1-9]|[1-4][0-9]|5[0-3])$
# ISO: YYYY-DDD: Ordinal date (day number from the year)	^(?:\d{4})-(00[1-9]|0[1-9][0-9]|[1-2][0-9]{2}|3[0-5][0-9]|36[0-6])$
# ISO: YYYYDDD: Ordinal date (day number from the year)	^(?:\d{4})(00[1-9]|0[1-9][0-9]|[1-2][0-9]{2}|3[0-5][0-9]|36[0-6])$
# ISO: YYYY: year	^(\d{4})$
# ISO: MM: month	^(0[1-9]|1[0-2])$
# ISO: DD: day	^(0[1-9]|[1-2][0-9]|3[01])$
# ISO: YYYY-MM-DDTHH:MM:SSZ: Date and Time Combined (UTC)	^(\d{4})-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01])T([01]\d|2[0-3]):([0-5]\d):([0-5]\d)Z$
# ISO: YYYY-MM-DDTHH:MM:SS±hh:mm: Date and Time Combined (with Timezone Offset)	^(\d{4})-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01])T([01]\d|2[0-3]):([0-5]\d):([0-5]\\d)([+-][01]\\d:[0-5]\d)$
# ISO: PnYnMnDTnHnMnS :durations e.g. P3Y6M4DT12H30M5S	^P(?!$)((\d+Y)|(\d+.\d+Y)$)?((\d+M)|(\d+.\d+M)$)?((\d+W)|(\d+.\d+W)$)?((\d+D)|(\d+.\d+D)$)?(T(?=\d)((\d+H)|(\d+.\d+H)$)?((\d+M)|(\d+.\d+M)$)?(\d+(.\d+S)?)?)?$
# ISO: HH:MM: hour, minutes in 24 hour notation	^([01]\d|2[0-3]):([0-5]\d)$
# ISO: HH:MM:SS: hour, minutes, seconds in 24 hour notation	^([01]\d|2[0-3]):([0-5]\d):([0-5]\d)$
# DD/MM/YYYY: day, month, year	^(0[1-9]|[12]\d|3[01])/(0[1-9]|1[0-2])/\d{4}$
# DD/MM/YY: day, month, year	^(0[1-9]|[12]\d|3[01])/(0[1-9]|1[0-2])/\d{2}$
# MM/DD/YYYY: month, day, year	^(0[1-9]|1[0-2])/(0[1-9]|[12]\d|3[01])/\d{4}$
# DDMMYYYY: day, month, year	^(0[1-9]|[12]\d|3[01])(0[1-9]|1[0-2])\d{4}$
# MMDDYYYY: month, day, year	^(0[1-9]|1[0-2])(0[1-9]|[12]\d|3[01])\d{4}$
# YYYYMMDD: year, month, day	^(\d{4})(0[1-9]|1[0-2])(0[1-9]|[1-2]\d|3[0-1])$
# HH:MM:SS: hour, minutes, seconds 12 hour notation AM/PM	^(0?[1-9]|1[0-2]):[0-5][0-9]:[0-5][0-9] ?[APMapm]{2}$
# H:MM or HH:MM: hour, minutes AM/PM	^(0?[1-9]|1[0-2]):[0-5][0-9] ?[APMapm]{2}$


import json
import optparse
import os
import sys
from collections import OrderedDict
import yaml
import csv
import re

###############################################################################

warnings = [];
locale_mapping = {}; # Converting OCA language terms to i18n

SCHEMA_CORE = r"""id: https://example.com/
name: ExampleSchema
title: Example Schema
description: An example schema description
version: 0.0.0
in_language: []
imports:
- linkml:types
prefixes:
  linkml: https://w3id.org/linkml/
classes:
  dh_interface:
    name: dh_interface
    description: A DataHarmonizer interface
slots: {}
enums: {}
types:
  WhitespaceMinimizedString:
    name: WhitespaceMinimizedString
    typeof: string
    description: "A string that has all whitespace trimmed off of beginning and end,
      and all internal whitespace segments reduced to single spaces. Whitespace includes
      #x9 (tab), #xA (linefeed), and #xD (carriage return)."
    base: str
    uri: xsd:token
settings:
  Title_Case: '((?<=\b)[^a-z\W]\w*?|[\W])+'
  UPPER_CASE: '[A-Z\W\d_]*'
  lower_case: '[a-z\W\d_]*'
"""

SCHEMA_SLOTS = [
	"class_name",
	"slot_group",
	"slot_uri",
	"name",
	"title",
	"range",
	"range_2",
	"identifier",
	"multivalued",
	"required",
	"recommended",
	"minimum_value",
	"maximum_value",
	"pattern",
	"structured_pattern",
	"description",
	"comments",
	"examples"
];

SCHEMA_ENUMS = [
	"title",
	"name",
	"meaning",
	"menu_1",
	"menu_2",
	"menu_3",
	"menu_4",
	"menu_5",
	"description"
]; # + language variants.

###############################################################################

def init_parser():
	parser = optparse.OptionParser()

	parser.add_option(
		"-i",
		"--input",
		dest="input_oca_file",
		default="oca_bundle.json",
		help="Provide an OCA json bundle file path and name to read.",
	)

	return parser.parse_args();

def save_tsv(file_name, headers, data):
	with open(file_name, 'w') as output_handle:
		csvwriter = csv.writer(output_handle,delimiter='\t', lineterminator='\n');
		csvwriter.writerow(headers);
		for row in data:
			if isinstance(row, OrderedDict):
				csvwriter.writerow(row.values());
			else: # row is an array or dict index
				csvwriter.writerow(data[row].values());

def localeLookup (language):
	# Ideally we're just matching i18n 2 or 3 letter language codes.
	match language:
		case "eng": return "en";
		case "fra":	return "fr";
		case _: return language;

def getLookup(overlay, oca_locale, key):
	for obj in oca_obj["bundle"]["overlays"][overlay]:
		if obj['language'] == oca_locale:
			match overlay:
				case "meta": 
					return obj[key];
				case "label":
					return obj["attribute_labels"][key];
				case "information":
					return obj["attribute_information"][key];
				case "entry":
					return obj["attribute_entries"][key]
	return "";

# For those schemas that have multilingual content
def addLocaleHeaders(headers, fields):
	# Ensure given slot or enum table has right language variation
	if len(locale_mapping) > 1:
		for field in fields:
			for locale in list(locale_mapping)[1:]:
				headers.append(field + "_" + locale);

# Make LinkML schema_core.yaml file which is then filled in with slots and 
# enumerations.
def writeSchemaCore():

	#with open("schema_core_template.yaml", 'r') as file:
	SCHEMA = yaml.safe_load(SCHEMA_CORE);

	SCHEMA["name"] = "DefaultSchemaName";
	# Is meta always in spec?
	if "meta" in oca_obj["bundle"]["overlays"]: 
		oca_meta = oca_metas; # FIRST meta is default
		if "name" in oca_meta:
			SCHEMA["title"] = oca_meta["name"]; # e.g. "Chicken gut health"
			# Schema name as PascalCase version of oca meta schema name.
			SCHEMA["name"] = "".join(filter(str.isalnum, oca_meta["name"].title()))

		SCHEMA["description"] = oca_meta["description"];

	# No other way to get/convey URI of schema at moment.
	SCHEMA["id"] = "https://example.com/" + SCHEMA["name"]; # The official URI of schema

	# Assume existence of label overlay means english if no language specified
	if "language" in oca_obj["bundle"]["overlays"]["label"][0]:
		for label_obj in oca_obj["bundle"]["overlays"]["label"]:
			locale = localeLookup(label_obj["language"]);
			SCHEMA["in_language"].append(locale)
			locale_mapping[locale] = label_obj["language"];

		if len(SCHEMA["in_language"]) == 1: # make it a single value
			SCHEMA["in_language"] = SCHEMA["in_language"][0];
	else:
		# Only do multiple languages if "language" parameter is present.
		SCHEMA["in_language"] = "en";

	SCHEMA["classes"][SCHEMA["name"]] =  {
	    'name': SCHEMA["name"],
    	'title': SCHEMA["title"] or SCHEMA["name"],
    	'description': SCHEMA["description"],
    	'is_a': 'dh_interface'
	}
	# ISSUE: Each class may have (meta) title and description fields translated
	# but we don't have a SCHEMA_CLASSES table to handle translations in 
	# tabular_to_schema.py, so can't communicate them.

	with open("schema_core.yaml", 'w') as output_handle:
		yaml.dump(SCHEMA, output_handle, sort_keys=False);

	return SCHEMA;

def writeSlots():
	# Ensure SCHEMA_SLOTS has language variation
	addLocaleHeaders(SCHEMA_SLOTS, ["slot_group","title","description","comments","examples"]);

	# start slots as an ordered dictionary {slot name:,...} of DH slot attributes.
	slots = OrderedDict([i, OrderedDict([i,""] for i in SCHEMA_SLOTS) ] for i in oca_attributes)

	for slot_name in oca_attributes:
		slot = slots[slot_name];
		slot['class_name'] = SCHEMA["name"];
		slot['name'] = slot_name;
		slot['title'] = oca_labels[slot_name];
		slot['range'] = oca_attributes[slot_name]; # ISSUE: Numeric
		slot['pattern'] = oca_formats[slot_name];
		slot['description'] = oca_informations[slot_name];

		# Range 2 gets any picklist for now.
		if slot_name in oca_entry_codes:
			slots[slot_name]['range_2'] = slot_name;

		# post-process range field:
		match slot['range']: # case sensitive?
			case "Text":
				slot['range'] = "WhitespaceMinimizedString" # or "string"
			case "Numeric":
				# ISSUE: if field is marked as an integer or decimal, then even
				# if regular expression validates, a test against integer or 
				# decimal format will INVALIDATE this slot.
				# Sniff whether it is integer or decimal. FUTURE: allow negatives?
				if re.search("^-?\[0-9\]\{\d+\}$", slot['pattern']):
					slot['range'] = "integer";
				else:
					slot['range'] = "decimal";

		# Need access to original oca language parameter, e.g. "eng"
		if len(locale_mapping) > 1:
			for locale in list(locale_mapping)[1:]:
				oca_locale = locale_mapping[locale];
				#slot['slot_group_'+locale]
				slot['title_'+locale] = getLookup("label", oca_locale, slot_name)
				slot['description_'+locale] = getLookup("information", oca_locale, slot_name)
				#slot['comments_'+locale]
				#slot['examples_'+locale]
	
	save_tsv("schema_slots.tsv", SCHEMA_SLOTS, slots);

def writeEnums():
	addLocaleHeaders(SCHEMA_ENUMS, ["title", "menu_1"]);
	enums = [];
	for enum_name in oca_entry_codes:
		defining_row = True;
		for enum_choice in oca_entry_codes[enum_name]:
			row = OrderedDict([i,""] for i in SCHEMA_ENUMS);
			if defining_row == True:
				row["name"] = enum_name;
				row["title"] = enum_name;
				defining_row = False;

			row["meaning"] = enum_choice;
			row["menu_1"] = oca_entry_labels[enum_name][enum_choice];

			if len(locale_mapping) > 1:
				for locale in list(locale_mapping)[1:]:
					oca_locale = locale_mapping[locale];
					row['menu_1_'+locale] = getLookup("entry", oca_locale, enum_choice)

			enums.append(row);

	save_tsv("schema_enums.tsv", SCHEMA_ENUMS, enums);


###############################################################################

options, args = init_parser();

# Load OCA schema bundle specification
if options.input_oca_file and os.path.isfile(options.input_oca_file):
    with open(options.input_oca_file, "r") as f:
        oca_obj = json.load(f);
else:
	os.exit("- [Input OCA bundle file is required]")

# In parsing input_oca_file, a few attributes are ignored for now in much of
# the structure:
# 	"type": [string absolute/relative URI], 
# 	"capture_base": hash, # Ignore
# ALSO, it is assumed that language variant objects all have the "default" 
# and consistent primary language as first entry.

oca_overlays = oca_obj["bundle"]["overlays"];

# Contains {schema.name,.description,.language} in array 
# Optional?
oca_metas = oca_overlays["meta"][0];

# oca_attributes contains slot.name and slot.datatype
oca_attributes = oca_obj["bundle"]["capture_base"]["attributes"];

# Contains slot.name and slot.pattern
oca_formats = oca_overlays["format"]["attribute_formats"];

# Contains {slot.title,.language} in array
oca_labels = oca_overlays["label"][0]["attribute_labels"];

# Contains {slot.name,.description,.language} in array 
# Optional?
oca_informations = oca_overlays["information"][0]["attribute_information"];

# Contains {"d": "M", "i": "M", "passed": "M"}  # "M" ?
oca_conformance = oca_overlays["conformance"]["attribute_conformance"];

# Contains [enumeration name]:[code,...]
oca_entry_codes = oca_overlays["entry_code"]["attribute_entry_codes"];

# Contains array of {enumeration.language,.attribute_entries} where 
# attribute_entries is dictionary of [enumeration name]: {code, label}
oca_entry_labels = oca_overlays["entry"][0]["attribute_entries"];

SCHEMA = writeSchemaCore();
writeSlots();
writeEnums();

if len(warnings):
	print ("\nWARNING: \n", "\n ".join(warnings));

sys.exit("Finished processing")

File: lib/OcaToLinkml.js
/* OcaToLinkml.js
 *
 * Convert an OCA (Overlay Capture Architecture) bundle JSON object to a
 * LinkML schema object suitable for YAML.stringify + loadSchemaYAML().
 *
 * Supports both the older "bundle" format and the newer "oca_package" format.
 * See: https://github.com/agrifooddatacanada/OCA_package_standard
 */

/**
 * Return true if the parsed JSON object looks like an OCA bundle.
 */
export function isOcaBundle(obj) {
  if (!obj || typeof obj !== 'object') return false;
  // Newer "oca_package" format: { type: "oca_package/...", oca_bundle: { bundle: ... } }
  if (typeof obj.type === 'string' && obj.type.split('/')[0] === 'oca_package') {
    return !!obj.oca_bundle?.bundle?.capture_base?.attributes;
  }
  // Older "bundle" format: { bundle: { capture_base: { attributes: {...} } } }
  return !!obj.bundle?.capture_base?.attributes;
}

/**
 * Convert an OCA bundle JSON object to a LinkML schema plain object.
 * The result can be serialised with YAML.stringify() and fed to loadSchemaYAML().
 */
export function ocaToLinkml(obj) {
  // ── Unpack oca_package wrapper if present ───────────────────────────────
  let bundle_obj;
  if (typeof obj.type === 'string' && obj.type.split('/')[0] === 'oca_package') {
    bundle_obj = obj.oca_bundle;
  } else {
    bundle_obj = obj;
  }

  const capture_base = bundle_obj.bundle.capture_base;
  const overlays     = bundle_obj.bundle.overlays || {};

  // ── Capture base ────────────────────────────────────────────────────────
  const attributes      = capture_base.attributes || {};
  const flagged_attrs   = new Set(capture_base.flagged_attributes || []);

  // ── Overlays ────────────────────────────────────────────────────────────
  const formats     = overlays.format?.attribute_formats || {};
  const cardinality = overlays.cardinality?.attr_cardinality || {};
  const conformance = overlays.conformance?.attribute_conformance || {};
  const entry_codes = overlays.entry_code?.attribute_entry_codes || {};

  // entry labels – use first (primary language) entry overlay
  const entry_arr    = _asArray(overlays.entry);
  const entry_labels = entry_arr[0]?.attribute_entries || {};

  // units
  let units = {};
  let metric_system = '';
  if (overlays.unit) {
    metric_system = overlays.unit.metric_system || '';
    units = overlays.unit.attribute_units || overlays.unit.attribute_unit || {};
  }

  // meta – use first entry as primary; derive schema name / title / description
  const meta_arr         = _asArray(overlays.meta);
  let schema_title       = 'UntitledSchema';
  let schema_description = '';
  let primary_locale     = 'en';
  if (meta_arr.length > 0) {
    const primary_meta = meta_arr[0];
    if (primary_meta.name)        schema_title       = primary_meta.name;
    if (primary_meta.description) schema_description = primary_meta.description.trim();
    if (primary_meta.language)    primary_locale     = _localeLookup(primary_meta.language);
  }
  const schema_name = _toPascalCase(schema_title) || 'UntitledSchema';

  // labels (attribute titles) – use first label overlay
  const label_arr = _asArray(overlays.label);
  const labels    = label_arr[0]?.attribute_labels || {};

  // information (attribute descriptions) – use first information overlay
  const info_arr     = _asArray(overlays.information);
  const informations = info_arr[0]?.attribute_information || {};

  // ── Build schema skeleton ────────────────────────────────────────────────
  const schema = {
    id:          `https://example.com/${schema_name}`,
    name:        schema_name,
    title:       schema_title,
    description: schema_description,
    version:     '0.0.1',
    in_language: primary_locale,
    imports:     ['linkml:types'],
    prefixes: {
      linkml: 'https://w3id.org/linkml/',
    },
    classes: {},
    slots:   {},
    enums:   {},
  };

  // ── Build class attributes ───────────────────────────────────────────────
  const class_attributes = {};

  for (const [slot_name, oca_type] of Object.entries(attributes)) {
    const pattern     = formats[slot_name];
    const type_result = _mapOcaType(oca_type, pattern);
    const slot_def    = {};

    if (labels[slot_name])       slot_def.title       = labels[slot_name];
    if (informations[slot_name]) slot_def.description = informations[slot_name];

    // Range: prefer enum (entry_codes) over primitive type
    slot_def.range = entry_codes[slot_name] ? slot_name : type_result.range;

    // Multivalued from Array[T] type
    if (type_result.multivalued) slot_def.multivalued = true;

    // Cardinality overlay (may also set multivalued)
    if (cardinality[slot_name]) {
      const card = _parseCardinality(cardinality[slot_name]);
      if (card.multivalued)          slot_def.multivalued          = true;
      if (card.minimum_cardinality !== undefined)
                                     slot_def.minimum_cardinality  = card.minimum_cardinality;
      if (card.maximum_cardinality !== undefined)
                                     slot_def.maximum_cardinality  = card.maximum_cardinality;
    }

    // Conformance
    if      (conformance[slot_name] === 'M') slot_def.required    = true;
    else if (conformance[slot_name] === 'O') slot_def.recommended = true;

    // Pattern (only for non-enum slots)
    if (pattern && !entry_codes[slot_name]) slot_def.pattern = pattern;

    // Unit (stored as { ucum_code } for addSlotRecord to read slot_obj.unit?.ucum_code)
    if (units[slot_name]) slot_def.unit = { ucum_code: units[slot_name] };

    // Identifying factor (flagged / confidentiality-sensitive attribute)
    if (flagged_attrs.has(slot_name)) {
      slot_def.annotations = { identifying_factor: { value: 'True' } };
    }

    class_attributes[slot_name] = slot_def;
  }

  // ── Main class ───────────────────────────────────────────────────────────
  schema.classes[schema_name] = {
    name:        schema_name,
    title:       schema_title,
    description: schema_description,
    attributes:  class_attributes,
  };

  // ── Container class (tree_root) ──────────────────────────────────────────
  schema.classes['Container'] = {
    name:       'Container',
    tree_root:  true,
    attributes: {
      [schema_name + 'Data']: {
        multivalued:     true,
        range:           schema_name,
        inlined_as_list: true,
      },
    },
  };

  // ── Enums ────────────────────────────────────────────────────────────────
  for (const [enum_name, codes] of Object.entries(entry_codes)) {
    const permissible_values = {};
    for (const code of codes) {
      const label = entry_labels[enum_name]?.[code];
      permissible_values[code] = label ? { title: label } : {};
    }
    schema.enums[enum_name] = {
      name:               enum_name,
      title:              enum_name,
      permissible_values,
    };
  }

  return schema;
}

// ── Private helpers ──────────────────────────────────────────────────────────

/** Normalise a value that may be an array or a single object to an array. */
function _asArray(val) {
  if (!val) return [];
  return Array.isArray(val) ? val : [val];
}

/**
 * Map OCA 3-letter ISO 639-2 codes to 2-letter i18n codes where known.
 * Unknown codes are returned unchanged.
 */
function _localeLookup(lang) {
  const MAP = {
    eng: 'en', fra: 'fr', deu: 'de', spa: 'es', por: 'pt',
    ita: 'it', nld: 'nl', pol: 'pl', rus: 'ru', zho: 'zh',
    jpn: 'ja', kor: 'ko', ara: 'ar', hin: 'hi',
  };
  return MAP[lang] || lang;
}

/**
 * Convert an arbitrary string to PascalCase, keeping only alphanumeric chars.
 * e.g. "Chicken gut health" → "ChickenGutHealth"
 */
function _toPascalCase(str) {
  return str
    .split(/[^a-zA-Z0-9]+/)
    .filter(Boolean)
    .map(word => word.charAt(0).toUpperCase() + word.slice(1))
    .join('');
}

/**
 * Map an OCA attribute type string to a LinkML range (and optional multivalued flag).
 * Handles Array[T] recursively.
 *
 * @param  {string}  oca_type  e.g. "Text", "Numeric", "Boolean", "Array[Text]"
 * @param  {string}  pattern   OCA format regex (used to sniff integer vs decimal)
 * @returns {{ range: string, multivalued?: true }}
 */
function _mapOcaType(oca_type, pattern) {
  // Array[T] → multivalued + inner type
  const array_match = oca_type.match(/^Array\[(.+)\]$/);
  if (array_match) {
    return { ..._mapOcaType(array_match[1], pattern), multivalued: true };
  }

  switch (oca_type) {
    case 'Text':
      return { range: 'string' };

    case 'Numeric':
      // Sniff integer: pattern consists of an optional minus, then [0-9]{n}
      if (pattern && /^-?\[0-9\]\{\d+\}$/.test(pattern)) {
        return { range: 'integer' };
      }
      return { range: 'decimal' };

    case 'Boolean':
      return { range: 'boolean' };

    case 'DateTime':
      return { range: 'datetime' };

    default:
      return { range: 'string' };
  }
}

/**
 * Parse an OCA cardinality string into cardinality fields.
 * Formats: "n", "n-", "-m", "n-m"
 *
 * @param  {string} card
 * @returns {{ minimum_cardinality?: number, maximum_cardinality?: number, multivalued?: true }}
 */
function _parseCardinality(card) {
  const result = {};
  if (!card) return result;

  if (card.includes('-')) {
    if (card[0] === '-') {
      // "-m": only maximum given
      const max = parseInt(card.slice(1), 10);
      if (!isNaN(max)) {
        result.maximum_cardinality = max;
        if (max > 1) result.multivalued = true;
      }
    } else if (card[card.length - 1] === '-') {
      // "n-": only minimum given, unbounded maximum
      const min = parseInt(card.slice(0, -1), 10);
      if (!isNaN(min)) {
        result.minimum_cardinality = min;
        result.multivalued = true;
      }
    } else {
      // "n-m": both given
      const [min_s, max_s] = card.split('-');
      const min = parseInt(min_s, 10);
      const max = parseInt(max_s, 10);
      if (!isNaN(min)) result.minimum_cardinality = min;
      if (!isNaN(max)) result.maximum_cardinality = max;
      if (!isNaN(max) && max > 1) result.multivalued = true;
    }
  } else {
    // "n": exact count
    const n = parseInt(card, 10);
    if (!isNaN(n)) {
      result.minimum_cardinality = n;
      result.maximum_cardinality = n;
      if (n > 1) result.multivalued = true;
    }
  }
  return result;
}
